### ISSUE TITLE
Date Format property in DatePicker not working on Windows with date Format

### ISSUE THREAD

**Issue by FrankPohl** [dotnet/maui#10805]
### Description

I have defined three DatePickers on a Xaml page.
``<DatePicker Date="02/25/2022" Format = "dd.mm.yyyy"/>``
``<DatePicker Date="02/25/2022" Format = "dd/mm/yyyy"/>``
``<DatePicker Date="02/25/2022" Format = "mm/dd/yyyy"/>``

This is shown like that on Windows. The french format "mm/dd/yyyy" is ocrrect but the German is showing only the day.
On Android the month is set to 0.

![image](https://user-images.githubusercontent.com/1506961/196950356-2237f126-f7f9-4f65-97e2-5698c0562567.png)
And like that on an Android Phone
![image](https://user-images.githubusercontent.com/1506961/196951358-cd1db95b-8f5f-43b1-8cef-770d5ff6dc24.png)

I had expected to see the controls showing three different date formats but this is not the case.
I also tried to set the date format in code behind using the formats from CultureInfo.CurrentCulture.DateTimeFormat but this did not work either.

A similar problem was discussed in issue #7972

### Steps to Reproduce

Create a new .NET MAUI app
Add a DatePicker to the app with the format set to the German format. The separator is a ". here.
``<DatePicker Date="02/25/2022" Format = "dd.mm.yyyy"/>``
``<DatePicker Date="02/25/2022" Format = "dd/mm/yyyy"/>``
``<DatePicker Date="02/25/2022" Format = "mm/dd/yyyy"/>``

### Link to public reproduction project repository

https://github.com/FrankPohl/DatePickerError

### Version with bug

6.0.400

### Last version that worked well

Unknown/Other

### Affected platforms

Android, Windows

### Affected platform versions

Windows SDK 10.0.19041.0 and Android Framework 29.0

### Did you find any workaround?

No

**Comment by MatthewB05** [comment]
Also, is there a reason why the DatePicker doesn't use the system region settings ?  It seems to be in US format and you have to manually set it to the value in ShortDateFormat, whereas my expectation would be that system formats are used except where the user specifically wants to specify a different format.

**Comment by MatthewB05** [comment]
In fact it's not just the date format that's broken, the entire control is ignoring regional settings **on windows** - the day/month names are in English and the week starting dates are all out by one day.   On Android these are both correct.

![image](https://user-images.githubusercontent.com/107632511/197720949-5b7c2ccf-9926-435a-879c-bb5b485a0124.png)

**Comment by FrankPohl** [comment]
You are right, it was shown at the expected position but in my example the month was given as 02 and 00 was shown for the month.

**Comment by MatthewB05** [comment]
I've had to revisit this problem but with more MAUI knowledge, I've come up with the following workaround (to go in e.g. App.xaml.cs).  I'm not sure if the bug is in MAUI or in WinUI (I actually lean towards the latter) but someone, somewhere, seems to have forgotten the existence of the world beyond the borders of the United States.

I'm not saying this is the best solution, but seems low-impact and universal in the application, and certainly covers my own requirements.

```
#if WINDOWS
            // default datepicker ignorant of date format and first day of week
            var userLanguages = global::Windows.System.UserProfile.GlobalizationPreferences.Languages;
            var dateFormat = new global::Windows.Globalization.DateTimeFormatting.DateTimeFormatter("shortdate", userLanguages);
            DatePickerHandler.Mapper.AppendToMapping("FixDateFormat", (handler, view) =>
            {
                handler.PlatformView.FirstDayOfWeek = (global::Windows.Globalization.DayOfWeek)(int)
                    System.Globalization.CultureInfo.CurrentCulture.DateTimeFormat.FirstDayOfWeek;
                handler.PlatformView.DateFormat = dateFormat.Patterns.FirstOrDefault() ?? "shortdate";
            });
#endif
```

**Comment by MatthewB05** [comment]
Though even then the day/month names are wrong, but easier to live with than the Sunday where Monday should be, and back-to-front dates, especially for an internal app.

**Comment by pocki** [comment]
The weekdays and months also wrong (english in my case) even if you set the language `handler.PlatformView.Language` in @MatthewB05 workaround.

**Comment by Zhanglirong-Winnie** [comment]
Verified this issue with Visual Studio 17.7.0 Preview 1.0. Can repro on Windows platform with sample project.
[DatePickerError-main.zip](https://github.com/dotnet/maui/files/11597195/DatePickerError-main.zip)
Windows:
![10805](https://github.com/dotnet/maui/assets/77434865/9c4070b0-8e0a-47ad-bcda-4766912e0d8f)
Android:
![10805 A](https://github.com/dotnet/maui/assets/77434865/92b4f409-152e-4cd3-8c1b-88571bc590b8)

**Comment by drossoft** [comment]
Any progress with this? DatePicker is essential in any multilingual application. The migration from Xamarin to MAUI is really frustrating.

**Comment by Brosten** [comment]
I'm also struggeling with date and timepickers on windows.
Sizing and positioning is a joke...

How do I "vote" for this issue? By subscribing? Giving thumbs up on the original post or in some other way?

**Comment by AndreKraemer** [comment]
For those struggling with localized dates in .NET MAUI Windows Apps, here’s a solution to get a properly formatted DatePicker:

1. In your Platforms\Windows\Package.appxmanifest, make the following changes:
change this:
```
  <Resources>
    <Resource Language="x-generate" />
  </Resources>
```
  
  to this:
  
```xml
  <Resources>
    <Resource Language="x-generate" />
    <Resource Language="en-US"/>
    <Resource Language="de-DE"/>	  
  </Resources>
```

This ensures that your date picker uses the system’s default date format and has localized strings for months and days. 

However, you might still encounter an issue with the first day of the week not being correct for your culture. To address this, modify the code in your MauiProgram.cs as follows (code taken from @MatthewB05's answer):

```
#if DEBUG
    		builder.Logging.AddDebug();
#endif

#if WINDOWS
      // use the correct first day of the week
      DatePickerHandler.Mapper.AppendToMapping("FixFirstDayOfWeek", (handler, view) =>
      {
        handler.PlatformView.FirstDayOfWeek = (Windows.Globalization.DayOfWeek)(int)
            System.Globalization.CultureInfo.CurrentCulture.DateTimeFormat.FirstDayOfWeek;
      });
#endif

      return builder.Build();
```

The result should be a DatePicker that adheres to your system’s settings and displays localized strings. Here’s a screenshot from my German system:
![image](https://github.com/dotnet/maui/assets/312671/45127c73-a312-43cf-bc34-af9e6d86801b)

**Comment by jfversluis** [comment]
@AndreKraemer thanks for adding this! Is this _the_ way to do it? Or is there anything we can add to .NET MAUI as well to make this a bit easier to work with?

**Comment by AndreKraemer** [comment]
This is a great question @jfversluis.

I've actually made an addition to the official documentation regarding localizing the datepicker, which you can find at [this link](https://learn.microsoft.com/en-us/dotnet/maui/user-interface/controls/datepicker?view=net-maui-8.0#localize-a-datepicker-on-windows).

The section I contributed outlines the process of adding culture codes to the Package.appxmanifest, which ensures localized texts and date formats. As WinUI3 necessitates the inclusion of all supported languages in the Package.appxmanifest (similar to how iOS apps require CFBundleLocalizations in info.plist), there might not be much we can add to .NET MAUI to simplify this further. We just have to hope that everybody reads the documentation 😁 

Regarding the mapper for adjusting the first day of the week in the calendar to align with the users culture: While it could feasibly be integrated into .NET MAUI, there's a consideration to be made about consistency. If the first day of the week is changed without updating the calendar format (which needs to be done manually), it could lead to confusion among users.

There is a similar scenario with the TimePicker, as discussed in [this GitHub issue](https://github.com/dotnet/maui/issues/12423). The issue revolves around the timepicker is alwys using the 12-hour format on windows, regardless of your regional settings. I dug a bit deeper and the reason is the default value for the TimePicker's format property being "t", which signifies the shorthand for the localized short time format. While this works seamlessly on Android and iOS, on Windows, "t" is disregarded, and the underlying code determines the format based on the presence of an uppercase H (like HH:mm).

To address this, I devised a small mapper to replace "t" with the actual short time string (such as hh:mm or HH:mm) from the system.

Now, here's where it gets interesting: While it might be tempting to incorporate this mapper directly into .NET MAUI, we must acknowledge that in a native WinUI3 app (without MAUI), the WinUI3 control itself doesn't respect the current locale and defaults to using the 12-hour format. Thus, there's a dilemma: Should the control behave differently in .NET MAUI and implicitly adopt the correct time format, or should it adhere to the user's expectations?

In both scenarios, integrating the mapper code into .NET MAUI might seem like the simpler option for most users. However, we need to be mindful that some individuals might find this confusing.

So, what's your perspective on this? Is there a general principle, such as ensuring that in MAUI, the control behaves akin to the native control if developers don't explicitly set any properties? Or should it align with user expectations?

**Comment by babucr** [comment]
For me this is working. Just add blank StringFormat in binding.
<DatePicker  Date ="{Binding FromDate, StringFormat}" Format="dd-MMM-yyyy" />

**Comment by mijenner** [comment]
@AndreKraemer and others, 

I tried to use the proposed workaround to Package.appxmanifest, and it does work on starting an app. However if I, during runtime, change culture, it doesn't re-read the settings, or maybe I'm just doing it wrong? [https://github.com/mijenner/MAUIappsFeatures](url)

Kind regards, 
Michael
